src/serac/physics/utilities/quadrature_data.hpp

```diff
@@ -0,0 +1,212 @@
+      // The only legal (portable, defined) way to do type punning in C++
+      std::memcpy(qfunc_ptr + i, data_ptr + j, sizeof(T));
+      j++;
+    }
+  }
+
+private:
+  /**
+   * @brief Storage layout of @p qfunc_ containing mesh and polynomial order info
+   */
+  detail::MaybeOwningPointer<mfem::QuadratureSpace> qspace_;
+  /**
+   * @brief Per-quadrature point data, stored as array of doubles for compatibility with Sidre
+   */
+  detail::MaybeOwningPointer<mfem::QuadratureFunction> qfunc_;
+
+  std::vector<T> data_;
+  /**
+   * @brief The stride of the array
+   */
+  static constexpr int stride_ = detail::ceil(sizeof(T) / static_cast<double>(sizeof(double)));
+};
+
+/**
+ * @brief "Dummy" specialization, intended to be used as sentinel
+ */
+template <>
+class QuadratureData<void> {
+};
+
+// A dummy global so that lvalue references can be bound to something of type QData<void>
+// FIXME: There's probably a cleaner way to do this, it's technically a non-const global
+// but it's not really mutable because no operations are defined for it
+extern QuadratureData<void> dummy_qdata;
+
+// Hijacks the "vdim" parameter (number of doubles per qpt) to allocate the correct amount of storage
+template <typename T>
+QuadratureData<T>::QuadratureData(mfem::Mesh& mesh, const int p, const bool alloc)
+    : qspace_(std::make_unique<mfem::QuadratureSpace>(&mesh, p + 1)),
+      // When left unallocated, the allocation can happen inside the datastore
+      // Use a raw pointer here when unallocated, lifetime will be managed by the DataCollection
+      qfunc_(alloc ? detail::MaybeOwningPointer<mfem::QuadratureFunction>{std::make_unique<mfem::QuadratureFunction>(
+                         &detail::retrieve(qspace_), stride_)}
+                   : detail::MaybeOwningPointer<mfem::QuadratureFunction>{new mfem::QuadratureFunction(
+                         &detail::retrieve(qspace_), nullptr, stride_)}),
+      data_(static_cast<std::size_t>(detail::retrieve(qfunc_).Size() / stride_))
+{
+  // To avoid violating C++'s strict aliasing rule we need to std::memcpy a default-constructed object
+  // See e.g. https://gist.github.com/shafik/848ae25ee209f698763cffee272a58f8
+  // also https://en.cppreference.com/w/cpp/numeric/bit_cast
+  // also https://chromium.googlesource.com/chromium/src/base/+/refs/heads/master/bit_cast.h
+  static_assert(std::is_default_constructible_v<T>, "Must be able to default-construct the stored type");
+  static_assert(std::is_trivially_copyable_v<T>, "Uses memcpy - requires trivial copies");
+}
+
+template <typename T>
+T& QuadratureData<T>::operator()(const int element_idx, const int q_idx)
+{
+  // A view into the quadrature point data
+  mfem::Vector view;
+  // Use the existing MFEM offset calculation logic instead of reimplementing it here
+  // Avoids making this code dependent on mfem::QuadratureSpace impl
+  detail::retrieve(qfunc_).GetElementValues(element_idx, q_idx, view);
```

Comment:
I've never explicitly specified an ordering on the quadrature points in mfem. Empirically, they seem to be ordered lexicographically (so that's what I do on our quadrature rules too), I don't know if there are other options or not.